File: services/customers-service/src/test/java/br/com/deltaglobalbank/customers/features/customers/updateCustomer/UpdateCustomerUseCaseTests.java
package br.com.deltaglobalbank.customers.features.customers.updateCustomer;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertThrows;
import static org.mockito.ArgumentMatchers.any;
import static org.mockito.ArgumentMatchers.eq;
import static org.mockito.Mockito.lenient;
import static org.mockito.Mockito.never;
import static org.mockito.Mockito.times;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.when;

import java.time.LocalDate;
import java.util.List;
import java.util.UUID;

import br.com.deltaglobalbank.customers.domain.audit.CustomerAuditRepository;
import br.com.deltaglobalbank.customers.domain.customer.Customer;
import br.com.deltaglobalbank.customers.domain.customer.CustomerNotFound;
import br.com.deltaglobalbank.customers.domain.customer.CustomerRepository;
import br.com.deltaglobalbank.customers.domain.customer.SubaggregateDoesNotBelongToCustomer;
import br.com.deltaglobalbank.customers.domain.customer.valueobjects.Address;
import br.com.deltaglobalbank.customers.domain.customer.valueobjects.BirthDate;
import br.com.deltaglobalbank.customers.domain.customer.valueobjects.Cpf;
import br.com.deltaglobalbank.customers.domain.customer.valueobjects.Email;
import br.com.deltaglobalbank.customers.domain.customer.valueobjects.FullName;
import br.com.deltaglobalbank.customers.domain.customer.valueobjects.Gender;
import br.com.deltaglobalbank.customers.domain.customer.valueobjects.MaritalStatus;
import br.com.deltaglobalbank.customers.domain.customer.valueobjects.MotherName;
import br.com.deltaglobalbank.customers.domain.customer.valueobjects.Phone;
import br.com.deltaglobalbank.customers.domain.shared.valueobjects.Uf;
import br.com.deltaglobalbank.customers.infrastructure.persistence.repositories.JpaDocumentIssuerRepository;
import com.github.f4b6a3.uuid.UuidCreator;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.mockito.Mock;
import org.mockito.junit.jupiter.MockitoExtension;

@ExtendWith(MockitoExtension.class)
class UpdateCustomerUseCaseTests {

    @Mock
    CustomerRepository customerRepository;
    @Mock
    CustomerAuditRepository customerAuditRepository;
    @Mock
    JpaDocumentIssuerRepository jpaDocumentIssuerRepository;

    private UpdateCustomerUseCase useCase;

    private final UUID tenantId = UUID.randomUUID();
    private final UUID actor = UUID.randomUUID();

    @BeforeEach
    void setUp() {
        useCase = new UpdateCustomerUseCase(customerRepository, customerAuditRepository, jpaDocumentIssuerRepository);
        lenient().when(jpaDocumentIssuerRepository.findAll()).thenReturn(List.of());
    }

    private Customer aCustomer() {
        return Customer.create(
            UuidCreator.getTimeOrderedEpoch(), tenantId,
            new Cpf("11144477735"), new FullName("João da Silva"),
            new BirthDate(LocalDate.of(1980, 5, 15)), Gender.MALE,
            new MotherName("Maria da Silva"), MaritalStatus.MARRIED,
            new Email("joao@exemplo.com"), new Phone("+5511999998888"),
            new Address("01310100", "Av Paulista", "São Paulo", new Uf("SP")),
            actor, List.of(), List.of()
        );
    }

    private UpdateCustomerRequest aRequest(List<BankAccountRequest> accounts) {
        return new UpdateCustomerRequest(
            "João Souza", LocalDate.of(1980, 5, 15), "male",
            "brasileira", "Maria da Silva", "married",
            "joao@novo.com", new PhoneRequest("+5511988887777"),
            new AddressRequest("01310100", "Av Paulista", null, null, null, "São Paulo", "SP", null),
            List.of(), accounts
        );
    }

    @Test
    void throwsCustomerNotFoundWhenMissing() {
        UUID id = UUID.randomUUID();
        when(customerRepository.findById(id, tenantId)).thenReturn(null);
        assertThrows(CustomerNotFound.class,
            () -> useCase.execute(new UpdateCustomerCommand(id, tenantId, actor, aRequest(List.of()))));
        verify(customerRepository, never()).save(any());
    }

    @Test
    void updatesCustomerAndPersists() {
        Customer customer = aCustomer();
        when(customerRepository.findById(customer.id(), tenantId)).thenReturn(customer);
        when(customerRepository.save(any())).thenReturn(customer);
        UpdateCustomerResponse response = useCase.execute(new UpdateCustomerCommand(
            customer.id(), tenantId, actor,
            aRequest(List.of(new BankAccountRequest(
                null, "341", "1234", "56789", null, "checking", "disbursement", true)))
        ));
        assertEquals("João Souza", response.fullName());
        assertEquals(1, response.bankAccounts().size());
        verify(customerRepository, times(1)).save(customer);
        verify(customerAuditRepository, times(1)).saveAll(any());
    }

    @Test
    void throwsWhenAccountIdDoesNotBelong() {
        Customer customer = aCustomer();
        when(customerRepository.findById(customer.id(), tenantId)).thenReturn(customer);
        assertThrows(SubaggregateDoesNotBelongToCustomer.class, () -> useCase.execute(new UpdateCustomerCommand(
            customer.id(), tenantId, actor,
            aRequest(List.of(new BankAccountRequest(
                UUID.randomUUID(), "341", "1234", "56789", null, "checking", "disbursement", true)))
        )));
        verify(customerRepository, never()).save(any());
    }
}

File: services/customers-service/src/test/kotlin/br/com/deltaglobalbank/customers/features/customers/updateCustomer/UpdateCustomerUseCaseTests.kt
package br.com.deltaglobalbank.customers.features.customers.updateCustomer

import br.com.deltaglobalbank.customers.domain.audit.CustomerAuditRepository
import br.com.deltaglobalbank.customers.domain.customer.Customer
import br.com.deltaglobalbank.customers.domain.customer.CustomerNotFound
import br.com.deltaglobalbank.customers.domain.customer.CustomerRepository
import br.com.deltaglobalbank.customers.domain.customer.SubaggregateDoesNotBelongToCustomer
import br.com.deltaglobalbank.customers.domain.customer.valueobjects.*
import br.com.deltaglobalbank.customers.domain.shared.valueobjects.Uf
import br.com.deltaglobalbank.customers.infrastructure.persistence.repositories.JpaDocumentIssuerRepository
import com.github.f4b6a3.uuid.UuidCreator
import io.mockk.every
import io.mockk.impl.annotations.MockK
import io.mockk.junit5.MockKExtension
import io.mockk.verify
import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.Assertions.assertThrows
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.extension.ExtendWith
import java.time.LocalDate
import java.util.UUID

@ExtendWith(MockKExtension::class)
class UpdateCustomerUseCaseTests {
    @MockK lateinit var customerRepository: CustomerRepository
    @MockK(relaxed = true) lateinit var customerAuditRepository: CustomerAuditRepository
    @MockK lateinit var jpaDocumentIssuerRepository: JpaDocumentIssuerRepository

    private lateinit var useCase: UpdateCustomerUseCase

    private val tenantId = UUID.randomUUID()
    private val actor = UUID.randomUUID()

    @BeforeEach fun setUp() {
        useCase = UpdateCustomerUseCase(customerRepository, customerAuditRepository, jpaDocumentIssuerRepository)
        every { jpaDocumentIssuerRepository.findAll() } returns emptyList()
    }

    private fun Customer() = Customer.create(
        id = UuidCreator.getTimeOrderedEpoch(), tenantId = tenantId,
        cpf = Cpf("11144477735"), fullName = FullName("João da Silva"),
        birthDate = BirthDate(LocalDate.of(1980, 5, 15)), gender = Gender.MALE,
        motherName = MotherName("Maria da Silva"), maritalStatus = MaritalStatus.MARRIED,
        email = Email("joao@exemplo.com"), phoneNumber = Phone("+5511999998888"),
        address = Address("01310100", "Av Paulista", "São Paulo", Uf("SP")),
        createdBy = actor, bankAccounts = emptyList(), documents = emptyList()
    )

    private fun Request(accounts: List<BankAccountRequest> = emptyList()) = UpdateCustomerRequest(
        fullName = "João Souza", birthDate = LocalDate.of(1980, 5, 15), gender = "male",
        nationality = "brasileira", motherName = "Maria da Silva", maritalStatus = "married",
        email = "joao@novo.com", phone = PhoneRequest("+5511988887777"),
        address = AddressRequest(cep = "01310100", street = "Av Paulista", city = "São Paulo", state = "SP"),
        documents = emptyList(), bankAccounts = accounts,
    )

    @Test fun `throws CustomerNotFound when missing`() {
        val id = UUID.randomUUID()
        every { customerRepository.findById(id, tenantId) } returns null
        assertThrows(CustomerNotFound::class.java) {
            useCase.execute(UpdateCustomerCommand(id, tenantId, actor, Request()))
        }
        verify(exactly = 0) { customerRepository.save(any()) }
    }

    @Test fun `updates customer and persists`() {
        val customer = Customer()
        every { customerRepository.findById(customer.id, tenantId) } returns customer
        every { customerRepository.save(any()) } returns customer
        val response = useCase.execute(UpdateCustomerCommand(
            customer.id, tenantId, actor,
            Request(listOf(BankAccountRequest(
                bankCode = "341", agency = "1234", accountNumber = "56789",
                accountType = "checking", purpose = "disbursement", isPrimary = true)))
        ))
        assertEquals("João Souza", response.fullName)
        assertEquals(1, response.bankAccounts.size)
        verify { customerRepository.save(customer) }
        verify { customerAuditRepository.saveAll(any()) }
    }

    @Test fun `throws when account id does not belong`() {
        val customer = Customer()
        every { customerRepository.findById(customer.id, tenantId) } returns customer
        assertThrows(SubaggregateDoesNotBelongToCustomer::class.java) {
            useCase.execute(UpdateCustomerCommand(
                customer.id, tenantId, actor,
                Request(listOf(BankAccountRequest(
                    id = UUID.randomUUID(), bankCode = "341", agency = "1234", accountNumber = "56789",
                    accountType = "checking", purpose = "disbursement", isPrimary = true)))
            ))
        }
        verify(exactly = 0) { customerRepository.save(any()) }
    }
}